src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs

```diff
@@ -0,0 +1,224 @@
+        /// </summary>
+        IgnoreCache = 2,
+    }
+
+    /// <summary>
+    /// Class for managing wrappers of COM IUnknown types.
+    /// </summary>
+    [CLSCompliant(false)]
+    public abstract partial class ComWrappers
+    {
+        /// <summary>
+        /// Interface type and pointer to targeted VTable.
+        /// </summary>
+        public struct ComInterfaceEntry
+        {
+            /// <summary>
+            /// Interface IID.
+            /// </summary>
+            public Guid IID;
+
+            /// <summary>
+            /// Must be pinned memory that is owned by the implementer of <see cref="ComputeVtables(object, CreateComInterfaceFlags, out int)"/>.
+            /// The memory must live as long as any COM interface consuming the table exists.
+            /// </summary>
+            public IntPtr Vtable;
+        }
+
+        /// <summary>
+        /// ABI for function dispatch of a COM interface.
+        /// </summary>
+        public struct ComInterfaceDispatch
+        {
+            public IntPtr vftbl;
+
+            /// <summary>
+            /// Given a <see cref="System.IntPtr"/> from a generated VTable, convert to the target type.
+            /// </summary>
+            /// <typeparam name="T">Desired type.</typeparam>
+            /// <param name="dispatchPtr">Pointer supplied to VTable function entry.</param>
+            /// <returns>Instance of type associated with dispatched function call.</returns>
+            public static unsafe T GetInstance<T>(ComInterfaceDispatch* dispatchPtr) where T : class
+            {
+                // See the CCW dispatch section in the runtime for details on the masking below.
+                const long DispatchThisPtrMask = ~0xfL;
+                var comInstance = *(ComInterfaceInstance**)(((long)dispatchPtr) & DispatchThisPtrMask);
+
+                return Unsafe.As<T>(GCHandle.InternalGet(comInstance->GcHandle));
+            }
+
+            private struct ComInterfaceInstance
+            {
+                public IntPtr GcHandle;
+            }
+        }
+
+        /// <summary>
+        /// Globally registered instance of the ComWrappers class.
+        /// </summary>
+        private static ComWrappers? s_globalInstance;
+
+        /// <summary>
+        /// Create an COM representation of the supplied object that can be passed to an non-managed environment.
+        /// </summary>
+        /// <param name="instance">A GC Handle to the managed object to expose outside the .NET runtime.</param>
+        /// <param name="flags">Flags used to configure the generated interface.</param>
+        /// <returns>The generated COM interface that can be passed outside the .NET runtime.</returns>
+        public IntPtr GetOrCreateComInterfaceForObject(object instance, CreateComInterfaceFlags flags)
+        {
+            if (instance == null)
+                throw new ArgumentNullException(nameof(instance));
+
```

Comment:
Do we want to have any argument validation of the flags here? E.g. if `TrackerSupport` flag is specified, this instance have to be registered as global?

Reply:
We need to allow specifying `TrackerSupport` even if it's not registered as global. That's the only way to efficiently do WinRT object activation.

More specifically, in the RCW case we need to be able to say "associate this pointer with this object" for a newly constructed object in that object's constructor. The way to do that in the ComWrappers API is to create a new ComWrappers-derived class and instance that just returns the given object and pass it the pointer.

Reply:
I was under impression that the typical WinRT app is going to have just one ComWrapper instance, but I see that it is not the case.

Do you believe that you are doing for the RCW case is right, or would it make sense to have a more direct API for it?

Reply:
Based on some of your other review comments, Aaron and I have chatted about using the idea of a context parameter for RCW creation to support the object activation scenario. A more direct API would definitely be nice, but I don’t believe it’s absolutely necessary.

Reply:
If we realize in future that we need a more direct API, what would it look like?

I would like to test whether it can be naturally if needed, or whether it would be unnatural addition.

Reply:
One option would be:

```
+ public object RegisterObjectForInterface(object obj, IntPtr externalComObject);
```

The API would add the mapping to the cache if it doesn't exist and then return the RCW for the COM object. In the activation scenario this would be the same object, but in the case of a race it returns the first one registered.

There's one example in CsWinRT where there can be a race: delegate marshalling. To avoid doing unneeded reflection (and because we can't determine the delegate's type without being told in advance), the delegate marshaling creates an RCW and then registers it instead of creating it within a ComWrappers instance. In the case of a race, we want to return only the registered RCW to the user.

Reply:
I like this API. Should we just add it now? Should it take flags like the Get api?

Reply:
I think we can add it now. And it should probably take flags so we can say "this needs tracker support".

Reply:
I've added flags to my proposed addition above.

Reply:
Can someone restate the need for this new API? It is basically the same as `GetOrCreateObjectForComInterface()` but always requires the creation of an `object`.

Reply:
This API is for the scenario where you have an IntPtr and an object created outside of the ComWrappers and you want to tell the runtime “this object is the RCW for this pointer”. RuntimeClass activation and Delegate marshaling are the two scenarios where CSWinRT would use these. If it’s not provided, then CsWinRT would need to make additional subclasses of ComWrappers to implement it.

Reply:
An alternative approach is to use the `ThreadStatic` approach mentioned below. The subclassing should happen anyways to optimize scenarios when look up isn't needed and the type is known. This API demonstrates an anti-pattern of always allocate and then possibly use. As opposed to the other API that defers allocation until required. The solution to the issue above is simply more code gen as opposed to creating an API that necessitates the anti-pattern of allocate first and then see if needed.

Reply:
My perspective here is to avoid the creation of anti-pattern APIs if the solution can be accomplished through more code-gen. If code-gen can't accomplish the scenario then we should have the conversation.

Reply:
After spending some more time with this I want to propose something that I think will address all concerns.

Instead of a new API that is yet another way to create an RCW, how about we change the `GetOrCreateObjectForComInterface()` to not take the `Type` instance as presently implemented but instead take an optional `object` which is the RCW. The semantics are the same, but if the `object` is provided the protected `CreateObject()` call will not be made since there is no need to create anything.

This keeps the semantics for creation clear - use `GetOrCreateObjectForComInterface()`, but allows for the annoying cases where the `object` was already created so there is no need to request creation.

Upon deeper reflection of this proposal my two biggest concerns are (1) How do consumers know to select this API versus the other? and (2) Making the API larger than needed. I get the problem and acknowledge it is real, I am simply not convinced this specific case warrants an entirely new function.